src/mdns/mdns.hpp

```diff
@@ -258,9 +258,15 @@ class Publisher : private NonCopyable
      *
      * @param[in] aType          The service type.
      * @param[in] aInstanceName  The service instance to subscribe, or empty to subscribe the service.
+     * @param[in] aKeepAlive     Whether the subscription should be kept alive to receive updates to the service
+     *                           instance. Note this only applies to avahi, as for avahi browse operation,
+     *                           instance updates will not trigger add/rm events.
+     *                           For dnssd browse operation, updates of instances will trigger add/rmv events.
      *
      */
-    virtual void SubscribeService(const std::string &aType, const std::string &aInstanceName) = 0;
+    virtual void SubscribeService(const std::string &aType,
+                                  const std::string &aInstanceName,
+                                  const bool         aKeepAlive) = 0;
```

Comment:
I'm thinking of if we are able to make 'keepalive' as the default behavior for all mDNS publishers.

As far as I know we have two use cases of subscribing a service:
1. Continuous browsing & resolving: TREL and SRPL. When a service updates its address, we want to get notified. 
 - mDNSResponder will send a remove event and an add event. The 'remove' event is optional from TREL/SRPL's perspective.
 - If we keep the Avahi resolver alive, Avahi will send an add event.

      The behaviors should be fine when we enable 'keep alive' for Avahi.

2. Discovery Proxy. In this case Discovery Proxy will unsubscribe the service after the service is resolved. The behaviors of Avahi and mDNSResponder are the same which is not impacted by 'keep alive'.

Reply:
Agree that we can make 'keepalive' as default behavior for all. When the caller of PublisherAvahi::SubscribeService doesn't want to get updates for the instances it should unsubscribe.

Will close this PR and open a new one.